File: DMC-E系列控制卡例程/c#/例2_连续运动/连续运动/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using csLTDMC;
using System.Threading;

namespace 连续运动
{
    public partial class Form1 : Form
    {
        private ushort usCardId = 0;
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            short sNum = LTDMC.dmc_board_init();//获取卡数量
            if (sNum <= 0 || sNum > 8)
            {
                MessageBox.Show("初始卡失败!", "出错");
            }
            ushort usNum = 0;
            ushort[] arrusCardList = new ushort[8];
            uint[] arrunCardTypes = new uint[8];
            short sRtn = LTDMC.dmc_get_CardInfList(ref usNum, arrunCardTypes, arrusCardList);
            if (sRtn != 0)
            {
                MessageBox.Show("获取卡信息失败!");
            }
            usCardId = arrusCardList[0];

            timer1.Start();
            // 使能对应轴    
            sRtn = LTDMC.nmc_set_axis_enable(usCardId, GetAxis());

        }

        private void Form1_FormClosed(object sender, FormClosedEventArgs e)
        {
            // 失能对应轴
            short sRtn = 0;
            sRtn = LTDMC.nmc_set_axis_disable(usCardId, GetAxis());
            LTDMC.dmc_board_close();
        }
        private ushort GetAxis()
        {
            return decimal.ToUInt16(nud_AxisId.Value);
        }
        private void btn_ChangeVel_Click(object sender, EventArgs e)
        {
            ushort usAxis = GetAxis(); //轴号
            double dNewVel = decimal.ToDouble(nud_NewVel.Value);                // 新的运行速度
            double dTaccDec = decimal.ToDouble(nud_TaccDec.Value);              //变速时间
            LTDMC.dmc_change_speed_unit(usCardId, usAxis, dNewVel, dTaccDec);  //在线变速
        }

        private void btn_Start_Click(object sender, EventArgs e)
        {
            ushort usAxis = GetAxis(); //轴号
            double dStartVel;           //起始速度
            double dMaxVel;             //运行速度
            double dTacc;               //加速时间
            double dTdec;               //减速时间
            double dStopVel;            //停止速度
            double dSpara;              //S段时间           
            ushort usDir;               //运动方向

            dStartVel = decimal.ToDouble(nud_StartVel.Value);
            dMaxVel = decimal.ToDouble(nud_MaxVel.Value);
            dTacc = decimal.ToDouble(nud_Tacc.Value);
            dTdec = decimal.ToDouble(nud_Tdec.Value);
            dStopVel = decimal.ToDouble(nud_StopVel.Value);
            dSpara = decimal.ToDouble(nud_Spara.Value);           
            if (radioButton_DirP.Checked)
            {
                usDir = 1;
            }
            else
            {
                usDir = 0;
            }
            LTDMC.dmc_set_profile_unit(usCardId, usAxis, dStartVel, dMaxVel, dTacc, dTdec, dStopVel);//设置速度参数
            LTDMC.dmc_set_s_profile(usCardId, usAxis, 0, dSpara);                                    //设置S段速度参数
            LTDMC.dmc_vmove(usCardId, usAxis, usDir);                                                 //定速运动
        }

        private void btn_Stop_Click(object sender, EventArgs e)
        {
            ushort usAxis = GetAxis();                             //轴号
            ushort usStopMode = 0;                                 //制动方式，0：减速停止，1：紧急停止
            double dTdec = decimal.ToDouble(nud_Tdec.Value);       //减速时间         
            LTDMC.dmc_set_dec_stop_time(usCardId, usAxis, dTdec);  //设置减速停止时间
            LTDMC.dmc_stop(usCardId, usAxis, usStopMode);
        }

        private void btn_SetPulseEquiv_Click(object sender, EventArgs e)
        {            
            ushort usAxis = GetAxis();
            double dEquiv = decimal.ToDouble(nud_PulseEquiv.Value);
            LTDMC.dmc_set_equiv(usCardId, usAxis, dEquiv);      //设置脉冲当量
        }

        private void btn_ClearPos_Click(object sender, EventArgs e)
        {
            ushort usAxis = GetAxis();                           //轴号
            double dPos = 0;                                     // 当前位置
            LTDMC.dmc_set_position_unit(usCardId, usAxis, dPos); //设置指定轴的当前指令位置值
        }

        private void btn_Esc_Click(object sender, EventArgs e)
        {
            this.Close();
        }

        private void timer1_Tick(object sender, EventArgs e)
        {
            ushort usAxis = GetAxis();  //轴号
            double dCurrentVel = 0;     //速度
            double dCmdPos = 0;         //指令位置
            double dEnPos = 0;          //编码器反馈位置

            LTDMC.dmc_read_current_speed_unit(usCardId, usAxis, ref dCurrentVel); // 读取轴当前速度
            tb_CurrentVel.Text = dCurrentVel.ToString();
            LTDMC.dmc_get_position_unit(usCardId, usAxis, ref dCmdPos);           //读取指定轴指令位置值
            tb_CurrentPos.Text = dCmdPos.ToString();
            LTDMC.dmc_get_encoder_unit(usCardId, usAxis, ref dEnPos);             //读取指定轴的编码器反馈值
            tb_Encoder.Text = dEnPos.ToString();

            if (LTDMC.dmc_check_done(usCardId, usAxis) == 0)                      //读取指定轴运动状态
            {
                tb_RunState.Text = "运行中";
            }
            else
            {
                tb_RunState.Text = "停止中";
            }


            // 读取指定轴状态机
            ushort usAxisStateMachine = 0;
            LTDMC.nmc_get_axis_state_machine(usCardId, usAxis, ref usAxisStateMachine);
            switch (usAxisStateMachine)
            {
                case 0: textBox_StateMachine.Text = "轴处于未启动状态";
                    textBox_StateMachine.BackColor = Color.Red;
                    break;
                case 1:
                    textBox_StateMachine.Text = "轴处于启动禁止状态";
                    textBox_StateMachine.BackColor = Color.Red;
                    break;
                case 2:
                    textBox_StateMachine.Text = "轴处于准备启动状态";
                    textBox_StateMachine.BackColor = Color.Red;
                    break;
                case 3:
                    textBox_StateMachine.Text = "轴处于启动状态";
                    textBox_StateMachine.BackColor = Color.Red;
                    break;
                case 4:
                    textBox_StateMachine.Text = "轴处于操作使能状态";
                    textBox_StateMachine.BackColor = Color.Green;
                    break;
                case 5:
                    textBox_StateMachine.Text = "轴处于停止状态";
                    textBox_StateMachine.BackColor = Color.Red;
                    break;
                case 6:
                    textBox_StateMachine.Text = "轴处于错误触发状态";
                    textBox_StateMachine.BackColor = Color.Red;
                    break;
                case 7:
                    textBox_StateMachine.Text = "轴处于错误状态";
                    textBox_StateMachine.BackColor = Color.Red;
                    break;
            };

            //读取总线状态
            ushort usErrorCode = 0;
            LTDMC.nmc_get_errcode(usAxis, 2, ref usErrorCode);
            if (usErrorCode == 0)
            {
                textBox_EthercatState.Text = "EtherCAT总线正常";
                textBox_EthercatState.BackColor = Color.Green;
            }
            else
            {
                textBox_EthercatState.Text = "EtherCAT总线出错";
                textBox_EthercatState.BackColor = Color.Red;
            }
        }

        private void button_HardwareReset_Click(object sender, EventArgs e)
        {
            richTextBox_Message.Text = "";
            richTextBox_Message.AppendText("请勿操作，总线卡硬件复位进行中……" + "\n");
            button_HardwareReset.Enabled = false;
            button_SoftwareReset.Enabled = false;
            //Application.DoEvents();
            LTDMC.dmc_board_reset();
            LTDMC.dmc_board_close();

            for (int i = 0; i < 15; i++)//总线卡硬件复位耗时15s左右
            {
                Application.DoEvents();
                Thread.Sleep(1000);
            }

            LTDMC.dmc_board_init();
            richTextBox_Message.AppendText("总线卡硬件复位完成,请确认总线状态");
            button_HardwareReset.Enabled = true;
            button_SoftwareReset.Enabled = true;
            button_HardwareReset.Focus();

        }

        private void button_SoftwareReset_Click(object sender, EventArgs e)
        {
            richTextBox_Message.Text = "";
            richTextBox_Message.AppendText("请勿操作，总线卡软件复位进行中……" + "\n");
            button_HardwareReset.Enabled = false;
            button_SoftwareReset.Enabled = false;
            //Application.DoEvents();
            LTDMC.dmc_soft_reset(usCardId);
            LTDMC.dmc_board_close();

            for (int i = 0; i < 15; i++)//总线卡软件复位耗时15s左右
            {
                Application.DoEvents();
                Thread.Sleep(1000);
            }
            LTDMC.dmc_board_init();
            richTextBox_Message.AppendText("总线卡软件复位完成,请确认总线状态");
            button_SoftwareReset.Enabled = true;
            button_HardwareReset.Enabled = true;
            button_HardwareReset.Focus();
        }


    }
}

File: DMC-E系列控制卡例程/VB.net/例2_连续运动/例2_连续运动/Form1.vb
Imports System.Threading

Public Class Form1
    Public g_usCardID As Short

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Dim sNum As Short
        Dim arr_nCardtypes(8) As UInteger
        Dim arr_usCardids(8) As UShort
        sNum = dmc_board_init() '获取卡数量

        If (sNum <= 0) Or (sNum > 8) Then             '正常的卡数在1- 8之间
            MsgBox("初始化LTDMC卡失败！", vbOKOnly, "出错")
        Else
            MsgBox("初始化LTDMC卡成功！", vbOKOnly, "OK")
        End If
        dmc_get_CardInfList(sNum, arr_nCardtypes, arr_usCardids) '
        g_usCardID = arr_usCardids(1) '
        Dim axis As UShort
        axis = GetAxis() '
        nmc_set_axis_enable(g_usCardID, axis) '
        Timer1.Start() '
    End Sub

    Private Function GetAxis() As UShort

        Return Decimal.ToUInt16(numericUpDown_Axis.Value) '

    End Function


    Private Sub Timer1_Tick(sender As Object, e As EventArgs) Handles Timer1.Tick
        Dim usAxis As UShort
        Dim dcurrent_V As Double
        Dim dpos As Double
        Dim np As Double
        Dim errcode As Int32
        Dim Axis_State_machine As Int32

        usAxis = GetAxis() '获取轴号
        dmc_read_current_speed_unit(g_usCardID, usAxis, dcurrent_V) '获取速度
        textBox1.Text = Str(dcurrent_V) '
        dmc_get_position_unit(g_usCardID, usAxis, dpos) '获取位置
        textBox2.Text = Str(dpos) '
        dmc_get_encoder_unit(g_usCardID, usAxis, np) ' 获取位置
        textBox3.Text = Str(np) '

        If (dmc_check_done(g_usCardID, usAxis) = 0) Then
            textBox4.BackColor = Color.Green
            textBox4.Text = "运行中" '
        Else
            textBox4.BackColor = Color.Red
            textBox4.Text = "停止中" '
        End If

        nmc_get_errcode(g_usCardID, 2, errcode)
        If (errcode = 0) Then
            TextBox6.Text = "EtherCAT总线正常"
            TextBox6.BackColor = Color.Green
        Else
            TextBox6.Text = "EtherCAT总线出错"
            TextBox6.BackColor = Color.Red
        End If

        nmc_get_axis_state_machine(g_usCardID, GetAxis(), Axis_State_machine)
        Select Case (Axis_State_machine) ' 读取指定轴状态机

            Case 0 : TextBox5.Text = "轴处于未启动状态"
                TextBox5.BackColor = Color.Red

            Case 1
                TextBox5.Text = "轴处于启动禁止状态"
                TextBox5.BackColor = Color.Red

            Case 2
                TextBox5.Text = "轴处于准备启动状态"
                TextBox5.BackColor = Color.Red

            Case 3
                TextBox5.Text = "轴处于启动状态"
                TextBox5.BackColor = Color.Red

            Case 4
                TextBox5.Text = "轴处于操作使能状态"
                TextBox5.BackColor = Color.Green

            Case 5
                TextBox5.Text = "轴处于停止状态"
                TextBox5.BackColor = Color.Red

            Case 6
                TextBox5.Text = "轴处于错误触发状态"
                TextBox5.BackColor = Color.Red

            Case 7
                TextBox5.Text = "轴处于错误状态"
                TextBox5.BackColor = Color.Red

        End Select


    End Sub


    Private Sub Form1_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing
        Timer1.Stop()
        dmc_board_close()
    End Sub

    Private Sub Btn_changevel_Click(sender As System.Object, e As System.EventArgs) Handles Btn_changevel.Click
        dmc_change_speed_unit(g_usCardID, GetAxis(), Decimal.ToDouble(numericUpDown_ChangeVel.Value), Decimal.ToDouble(numericUpDown_ChangeTime.Value)) '在线变速
    End Sub



    Private Sub Btn_StartMove_Click(sender As System.Object, e As System.EventArgs) Handles Btn_StartMove.Click
        Dim usAxis As UShort
        Dim usDir As UShort
        usAxis = GetAxis() ' 获取轴号
        dmc_set_profile_unit(g_usCardID,
                usAxis,
                Decimal.ToDouble(numericUpDown_StartSpeed.Value),
                Decimal.ToDouble(numericUpDown_MaxVel.Value),
                Decimal.ToDouble(numericUpDown_Tacc.Value),
                Decimal.ToDouble(numericUpDown_Tdcc.Value),
                Decimal.ToDouble(numericUpDown_StopTime.Value)) '
        dmc_set_s_profile(g_usCardID, usAxis, 0, Decimal.ToDouble(numericUpDown_STime.Value)) ' 设置运动参数
        If radioButton8.Checked Then
            usDir = 1 '  获取方向
        Else
            usDir = 0
        End If
        dmc_vmove(g_usCardID, usAxis, usDir) '连续运动
    End Sub

    Private Sub Btn_PosRest_Click(sender As System.Object, e As System.EventArgs) Handles Btn_PosRest.Click
        dmc_set_position_unit(g_usCardID, GetAxis(), 0) '位置设置成0
    End Sub

    Private Sub Btn_SetEquiv_Click(sender As System.Object, e As System.EventArgs) Handles Btn_SetEquiv.Click
        dmc_set_equiv(g_usCardID, GetAxis(), Decimal.ToDouble(numericUpDown_Equiv.Value)) '设置脉冲当量
    End Sub

    Private Sub Btn_Close_Click(sender As System.Object, e As System.EventArgs) Handles Btn_Close.Click
        nmc_set_axis_disable(g_usCardID, 0)
        nmc_set_axis_disable(g_usCardID, 1)
        nmc_set_axis_disable(g_usCardID, 2)
        Me.Close() ' 关闭程序
    End Sub

    Private Sub Btn_StopVel_Click(sender As System.Object, e As System.EventArgs) Handles Btn_StopVel.Click
        Dim usaxis As UShort
        usaxis = GetAxis() '
        dmc_stop(g_usCardID, usaxis, 1)
    End Sub

    Private Sub Form1_FormClosed(sender As System.Object, e As System.Windows.Forms.FormClosedEventArgs) Handles MyBase.FormClosed
        dmc_board_close()
        Dim axis As UShort
        axis = GetAxis() '
        nmc_set_axis_disable(g_usCardID, axis) '
    End Sub

    Private Sub Btn_HardReset_Click(sender As System.Object, e As System.EventArgs) Handles Btn_HardReset.Click
        RichTextBox1.Text = ""
        RichTextBox1.AppendText("请勿操作，总线卡硬件复位进行中……")
        Btn_HardReset.Enabled = False
        Btn_SoftReset.Enabled = False

        dmc_board_reset()
        dmc_board_close()

        For i As Integer = 0 To 15 Step 1 '总线卡硬件复位耗时15s左右
            Application.DoEvents()
            Thread.Sleep(1000)
        Next

        dmc_board_init()
        RichTextBox1.AppendText("总线卡硬件复位完成,请确认总线状态")
        Btn_HardReset.Enabled = True
        Btn_SoftReset.Enabled = True
        Btn_HardReset.Focus()
    End Sub

    Private Sub Btn_SoftReset_Click(sender As System.Object, e As System.EventArgs) Handles Btn_SoftReset.Click
        RichTextBox1.Text = ""
        RichTextBox1.AppendText("请勿操作，总线卡软件复位进行中……")
        Btn_HardReset.Enabled = False
        Btn_SoftReset.Enabled = False

        dmc_soft_reset(g_usCardID)
        dmc_board_close()

        For i As Integer = 0 To 15 Step 1 '总线卡硬件复位耗时15s左右
            Application.DoEvents()
            Thread.Sleep(1000)
        Next

        dmc_board_init()
        RichTextBox1.AppendText("总线卡软件复位完成,请确认总线状态")
        Btn_HardReset.Enabled = True
        Btn_SoftReset.Enabled = True
        Btn_SoftReset.Focus()
    End Sub
End Class
